contributors/design-proposals/network/service-topology-key.md

```diff
@@ -0,0 +1,193 @@
+* For a sharded service that keeps per-node local information in each shard.
+* Authenticating proxies such as [aws-es-proxy](https://github.com/kopeio/aws-es-proxy).
+* In container identity wg, being able to give daemonset pods a unique identity per host is on the 2018 plan, and ensuring local pods can communicate to local node services securely is a key goal there. -- from @smarterclayton
+* Regional data costs in multi-AZ setup - for instance, in AWS, with a multi-AZ setup, half of the traffic will switch AZ, incurring regional data Transfer costs, whereas if something was local, it wouldn't hit the network.
+* Users want a priority-based topology aware routing service. For example, want to visit a "same-zone" service,  of course, a "same-host" backend is more preferred.
+
+## Background
+
+It has been a very often requested feature because people need to create a service which connects to local backend Pods, the so called "only-local" service. In this way, they can reduce network latency, improve security, save money and so on. However, because topology is arbitrary, zone, region, rack, generator, whatever, who knows? We should allow arbitrary locality.
+
+`ExternalTrafficPolicy` was added in v1.4, but only for NodePort and external LB traffic. NodeName was added to `EndpointAddress` to allow kube-proxy to filter local endpoints for various future purposes.
+
+Based on our experience of advanced routing setup and recent demo of enabling this in Kubernetes service, this document would like to introduce a more generic way to enable "local-only service".
+
+## Proposal
+
+This proposal builds off of earlier requests to [use local pods only for kube-proxy loadbalancing](https://github.com/kubernetes/kubernetes/issues/7433) and [node-local service proposal](https://github.com/kubernetes/kubernetes/pull/28637). But, this document proposes that not only the particular "node-local" user case should be taken care, but also a more generic way should be figured out.
+
+## Implementation details
+
+### API changes
+
+#### Service API changes
+
+As with scheduling topology, there are currently two types of topology, called `Required` and `Preferred` which denote “hard” vs. “soft” requirements. 
+
+Similarly, we introduce `Required` and `Preferred` in service topology.
+
+An example of `required` topology would be “only visit service backends in the same zone”.
+
+An example of `preferred` topology would be “prefer to visit service backends in the same host when multiple endpoints are in the same zone”. Then "preferred" endpoints may have more routing probabilities.
+
+```go
+type Service struct {
+  Spec:
+    // If the topology requirements specified by this field are not met, kube-proxy will
+    // not pick endpoints for the service.
+    // ordering doesn't matter, it's an "AND" semantic in Required array.
+    // +optional
+    Required []TopologyTerm
+    // The kube-proxy will prefer to pick endpoints that satisfy
```

Comment:
what is the expected behavior for things that contact services from a location "outside" the specified topology, not via kube-proxy? for example, proxying to services via the API server (/api/v1/namespaces/ns1/services/services1/proxy) or resolving endpoints for aggregated API servers, admission webhooks, etc? is it a violation of the service spec to ignore a required topology term?

Reply:
> what is the expected behavior for things that contact services from a location "outside" the specified topology, not via kube-proxy?

I find the feature "ExternalTrafficLocalOnly" is not supported for apiserver proxy. Because this proposal is an extension of "ExternalTrafficLocalOnly", I don't have strong confidence of knowing how to implement it in apiserver proxy.